### DISCUSSION TITLE
Error "generateStaticParams()" so it cannot be used with "output: export" config.

### DISCUSSION THREAD

**Discussion by PedrooMota** [vercel/next.js#56731] [Help]
### Summary

I can't build my project, because it gives the following error:

` Page "/api/auth/[...nextauth]" is missing "generateStaticParams()" so it cannot be used with "output: export" config.`

Even using `use client` or `export const dynamic = 'force-static'` in route.ts, my build still gives this error

**I checked other discussions, but none of them solved my problem. If anyone knows what to do, I would be grateful.**

### versions used were:

1. next: 13.5.4
2. next-auth: 4.23.2

![image](https://github.com/vercel/next.js/assets/83295376/1358bdf2-9cce-4840-b7f3-1a221a57af32)

**Comment by icyJoseph** [comment]
Right, so if I make the assumption that you are using `/api` to create Route Handlers, and also use `export` mode, then I have to tell you that you are trying to combine two different things.

`export` mode builds a static HTML site, only HTML+JS+CSS to be sent by a "static files server" down to a browser, but the "static file server" doesn't do any other handling, in fact, you have to provide yourself, usually hosting providers do that.

To use Route handlers, you need a Node.js runtime running the Next.js Server (with next start). This is why the compiler wants you to define `generateStaticParams` so that it can try to build those during next build, and put them into the static HTML output.

**Reply by PedrooMota** [reply]
Before deploying the application to Vercel, I am using github pages. It necessarily requires the use of ouput: "export". 

So, I ended up trying to use it, but apparently it requires my route.ts file to be static. I'm still only using the credentials provider. With that, I would like to know if there is a way to deploy without this problem, or if I deploy on Vercel, I will not have the following problem.

To be honest, I'm a little lost on this issue.

```
import { NextAuthOptions } from "next-auth";
import NextAuth from "next-auth/next";
import CredentialsProvider from 'next-auth/providers/credentials';
import { URL_API } from '../../../../utils/constants'

const authOptions: NextAuthOptions = {
    providers: [
        CredentialsProvider({
            name: 'credentials',
            
            credentials: {
                email: {},
                password: {}
            },

            async authorize ( credentials: any ) {
                const payload = {
                    email: credentials.email,
                    password: credentials.password,
                };

                const res = await fetch(`${URL_API}/clients/auth/login`, {
                    method: 'POST',
                    body: JSON.stringify(payload),
                    headers: {
                        'Content-Type': 'application/json',
                    },
                });

                const user = await res.json();
                if (user !== null && res.status === 200) {
                    let { id, name, email, cpf, contact } = user.data;
                    let { AccessToken, RefreshToken, IdToken } = user;

                    return {
                        AccessToken,
                        IdToken,
                        RefreshToken,
                        id,
                        name,
                        email,
                        cpf,
                        contact,
                    };
                } else {
                    return null;
                }
            }
        }),
    ],
    callbacks: {
        async jwt({ token, user }: any) {
            if (user) {
                const isSignIn = user !== undefined;
                const DataAtualSeconds = Math.floor(Date.now() / 1000);
                const ExpirationTokenSeconds = Math.floor(3 * 24 * 60 * 60);

                if (isSignIn) {
                    if (user === null) {
                        return await Promise.resolve({});
                    }
                    token.AccessToken = user.AccessToken;
                    token.IdToken = user.IdToken;
                    token.RefreshToken = user.RefreshToken;
                    token.id = user.id;
                    token.name = user.name;
                    token.email = user.email;
                    token.cpf = user.cpf;
                    token.contact = user.contact;
                    token.expiration = Math.floor(DataAtualSeconds + ExpirationTokenSeconds);
                }else {
                    if (token.expiration === null) return await Promise.resolve({});
                    if (DataAtualSeconds > token.expiration) return await Promise.resolve({});
                }
            }
            return token;
        },

        async session({ session, token }: any) {
            if (token === null){
                return null;
            }
            session.AccessToken = token.AccessToken;
            session.IdToken = token.IdToken;
            session.RefreshToken = token.RefreshToken;
            session.user = {
                id: token.id,
                name: token.name,
                email: token.email,
                cpf: token.cpf,
                contact: token.contact
            };
            return {...session}
        },
    },

    pages: {
        signIn: '/signin',
        signOut: '/',
        newUser: '/dashboard'
    },

    session: {
        maxAge: 3 * 24 * 60 * 60, // 30 dias de token
    },

    secret: process.env.NEXTAUTH_SECRET,
    debug: process.env.NODE_ENV === "development"
};

const handler = NextAuth(authOptions);

export { handler as GET, handler as POST}
```

**Reply by icyJoseph** [reply]
Yeah but the issue again is that you are mixing up concerns here.

A GitHub pages deploy is but a static HTML server, it doesn't do any API handling for you. It has no server whatsoever.

You have a POST request on that file, and that'll need a server to handle the request. And again there's no such thing in a static HTML deploy.

**Reply by PedrooMota** [reply]
I understood. The issue is that locally the application works without using the output: "export", but when deploying to github pages, it requires that you use this command, and that the route.ts file is static. I would like to know if there is a way to deploy without this problem occurring, or if I use another method (Vercel), I can deploy smoothly.

**Reply by jeffreyschultz** [reply]
I need this ability because we have two deployment footprints. One is hosted within a container that is running `nextjs` server where `next-auth` is handling authentication, and another that is being hosted within `nginx`, where `nginx` is handling authentication as part of a more complex deployment.

Is there a way to turn off the API Routes for `mode: export` because when that is set, I don't need the API route functionality? All of my APIs are deployed separately, and this was purely for securing the webapp when serving with `nextjs`.

I guess the question would ultimately be, do I need to be using the App Router, and just go back to the Page Router?

**Reply by airtonix** [reply]
> It has no server whatsoever.

clarification: 

It has no **runtime** server, but it definitely is still a server... just one that only serves static assets matched from disk (and it has some restrictive ideals about what file extensions get served as what mimetype)

**Reply by airtonix** [reply]
@jeffreyschultz  In your case I'd modify your build step so it deletes the `app/api` directory before making a build

**Reply by Tosinkoa** [reply]
@icyJoseph Pls what can I do in a case where I have [...nextauth] and I am trying to do a build. I've added all my credential logic to my app using next auth, and I need to know if I would have to remove next auth, or there is a workaround to this.

**Reply by RjMatic** [reply]
i also encounter this error, i'm about to change my career

**Comment by jsuryakt** [comment]
**TL;DR**: The error fires when `generateStaticParams()` returns `[]`, not when it's actually missing. The fix is to ensure it always returns at least one param.

I hit this on Next.js 16.2.4 with `output: "export"` and Turbopack. My `generateStaticParams` fetched from a remote API that was returning HTTP 500, so the catch block returned `[]`. The error message led me down a hours-long rabbit hole inspecting compiled Turbopack chunks, thinking the export was being tree-shaken.

**What actually happens:**

```
generateStaticParams returns [] → Next.js treats it as "missing" → misleading error
```

**Minimal fix — always return at least one param:**

```ts
export async function generateStaticParams() {
  try {
    const res = await fetch(`${process.env.API_URL}/posts`);
    if (!res.ok) return [{ slug: "placeholder" }];
    const posts: { slug: string }[] = await res.json();
    if (posts.length === 0) return [{ slug: "placeholder" }];
    return posts.map((p) => ({ slug: p.slug }));
  } catch {
    return [{ slug: "placeholder" }];
  }
}
```

The page component handles unknown slugs with `notFound()`, so `/blog/placeholder` just renders a 404 — harmless.

**Suggestion**: The error message should differentiate between "function not exported" vs "function returned empty array." Something like:

```
Error: "generateStaticParams()" for page "/blog/[slug]" returned an empty array.
With "output: export", at least one param must be generated. If this route has
no content yet, return a placeholder value or remove the dynamic route.
```

**Comment by DaCameraGirl** [comment]
This error is a fundamental incompatibility between `output: 'export'` and any **Route Handler** (API route) in Next.js App Router. Here's a clear explanation and all the solutions.

## Root Cause

`output: 'export'` produces a completely static HTML/CSS/JS site — there is **no Node.js server** at runtime. Route Handlers (files named `route.ts` in `app/api/...`) require a running Node.js server to handle requests. These two things are mutually exclusive.

Next Auth uses a catch-all route handler at `/api/auth/[...nextauth]/route.ts`, which requires server execution. Static exports can't include this.

## Solution 1: Remove/Exclude the API Routes from the Export

If your API routes (like next-auth) live on a separate server or you're handling auth differently in production, exclude them from the build:

```ts
// next.config.ts
const nextConfig = {
  output: 'export',
}
```

Then **delete or move** `app/api/auth/[...nextauth]/route.ts` out of the project before running `next build`. The static export won't need it if auth is handled externally.

## Solution 2: Use a Separate Server for Auth (Most Common Pattern)

For a fully static site that still needs authentication, handle auth at a different layer:

- **Auth.js v5 (next-auth v5)** supports JWT-only mode without a server — tokens are stored in cookies and validated client-side or at edge middleware level
- Deploy your auth server separately (e.g., a standalone Node.js API or an auth provider like Auth0, Clerk, Supabase Auth)
- Use **Middleware** in Next.js to protect routes — middleware runs at the edge and works with static exports:

```ts
// middleware.ts
import { auth } from 'next-auth'

export default auth((req) => {
  if (!req.auth && req.nextUrl.pathname !== '/login') {
    return Response.redirect(new URL('/login', req.url))
  }
})

export const config = {
  matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)'],
}
```

> Note: Middleware works with `output: 'export'` when deployed to a platform that supports edge functions (Vercel, Cloudflare Pages). It does NOT work for local file-based exports served via a CDN.

## Solution 3: Don't Use `output: 'export'` — Use Server Deployment

If you need both:
- Next Auth (server-side sessions)
- Dynamic API routes

...then `output: 'export'` is the wrong choice. Remove it and deploy with `next start` or any Node.js-compatible platform (Vercel, Railway, Fly.io, etc.):

```ts
// next.config.ts
const nextConfig = {
  // Remove 'output: export'
  // output: 'export', <-- delete this line
}
```

## Solution 4: `generateStaticParams` Returns Empty Array (Different Bug)

If your error is on a **non-API dynamic route** (not next-auth), and `generateStaticParams` returns an empty array at build time, Next.js can't statically generate those pages. Fix:

```ts
export async function generateStaticParams() {
  const results = await fetchSlugs()
  
  // Guard: never return empty array with output: export
  if (results.length === 0) {
    return [{ slug: 'placeholder' }]
  }
  
  return results.map((item) => ({ slug: item.slug }))
}
```

## Quick Decision Tree

```
Do you need next-auth server sessions?
├── Yes → Remove output: 'export', deploy to Node.js server
└── No  → Can you use JWT cookies + middleware auth?
    ├── Yes → Use auth middleware only, remove /api/auth route
    └── No  → Use external auth provider (Clerk, Auth0, Supabase)
```

For most apps using `output: 'export'` + auth, **Solution 2 with next-auth v5 JWT mode + middleware** is the cleanest path forward.

**Comment by reziferdiansyah** [comment]
`This error happens because `output: "export"` generates a fully static site.

If you use a dynamic route like:

```txt
app/blog/[slug]/page.tsx`

Next.js must know all possible slug values at build time. You can fix it by adding generateStaticParams():
`export async function generateStaticParams() {
  const posts = await getPosts()

  return posts.map((post) => ({
    slug: post.slug,
  }))
}`

Example:
`return [
  { slug: "post-1" },
  { slug: "post-2" },
  { slug: "post-3" },
]`

But if your route depends on data that is only available at runtime, then output: "export" is not suitable.

In that case, remove this from next.config.js:
`output: "export"`

and deploy using a Next.js server runtime such as Vercel, Node.js, Docker, or AWS EC2.

Summary:

Use output: "export" only when all dynamic params are known at build time.
Use generateStaticParams() for dynamic routes.
Remove output: "export" if you need runtime dynamic routes.

Static export cannot generate unknown dynamic routes at request time.

**Comment by SukkaW** [comment]
I have created https://github.com/vercel/next.js/pull/95388 to fix the misleading error message here.
